### PR TITLE
SPIR-V handling on different OpenCL devices version

### LINKED ISSUE #716: SPIR-V handling on different OpenCL devices version
I have proposal how to handle SPIR-V compilation on different OpenCL devices.

### PR DESCRIPTION
**Pull request by gwawiork**
Fixes #716

### CODE REVIEW

File: test_conformance/math_brute_force/main.cpp [hunk 1]
```diff
@@ -1161,7 +1161,11 @@ test_status InitCL( cl_device_id device )
     }
 
     gDevice = device;
-
+    test_status spirv_status;
+    spirv_status = check_spirv_compilation_readiness(device);
```
Comment (kpet) [inline test_conformance/math_brute_force/main.cpp]: Is there any reason we can't do that for all suites in the harness code?

File: .gitignore [hunk 2]
```diff
@@ -1,4 +1,9 @@
...
 build/
 build_lnx/
-build_win/
\ No newline at end of file
+build_win/
+/__intel/*
+/binaries_to_zip/*
+/lib*
+/test_conformance/intel/*
+/test_conformance/gl/GL/*
```
Comment (kpet) [inline .gitignore]: I'm not sure all of these really belong upstream.

File: test_conformance/math_brute_force/main.cpp (hunk 1 above)
Comment (gwawiork) [reply test_conformance/math_brute_force/main.cpp]: Ok , done

File: .gitignore (hunk 2 above)
Comment (gwawiork) [reply .gitignore]: Yes, my mistake. Thanks.

**Comment by aarongreig** [comment]
with these changes can the minimum version [here](https://github.com/KhronosGroup/OpenCL-CTS/blob/master/test_conformance/spirv_new/main.cpp#L84) be lowered to (1, 2)?

File: test_conformance/spirv_new/main.cpp [hunk 3]
```diff
@@ -154,26 +154,50 @@ int get_program_with_il(clProgramWrapper &prog,
...
-    prog = clCreateProgramWithIL(context, buffer, file_bytes, &err);
-    SPIRV_CHECK_ERROR(err, "Failed to create program with clCreateProgramWithIL");
+    if (gCoreILProgram) {
+        prog = clCreateProgramWithIL(context, buffer, file_bytes, &err);
+        SPIRV_CHECK_ERROR(err, "Failed to create program with clCreateProgramWithIL");
+    }
+    else {
+        cl_platform_id platform;
+        err = clGetDeviceInfo(deviceID, CL_DEVICE_PLATFORM, sizeof(cl_platform_id), &platform, NULL);
+        print_error(err, "clGetDeviceInfo for CL_DEVICE_PLATFORM failed");
```
Comment (aarongreig) [inline test_conformance/spirv_new/main.cpp]: missing `if (err)`

**Comment by aarongreig** [comment]
I think the [linkage](https://github.com/KhronosGroup/OpenCL-CTS/blob/master/test_conformance/spirv_new/test_linkage.cpp#L36) and [opaque](https://github.com/KhronosGroup/OpenCL-CTS/blob/master/test_conformance/spirv_new/test_op_opaque.cpp#L31) tests will also need updating

**Comment by gwawiork** [comment]
> with these changes can the minimum version [here](https://github.com/KhronosGroup/OpenCL-CTS/blob/master/test_conformance/spirv_new/main.cpp#L84) be lowered to (1, 2)?

Are you able to run these tests to verify if they pass on OpenCL 1.2 ? I am not.

**Comment by aarongreig** [comment]
yes, I can confirm that they either pass or skip appropriately

File: test_common/harness/testHarness.cpp [hunk 4]
```diff
@@ -915,6 +929,128 @@ Version get_device_cl_version(cl_device_id device)
...
+
+    if (strstr(str.data(), "SPIR-V") == NULL) {
+        log_info("This device does not support SPIR-V offline compilation.\n");
+        return false;
+    } else {
+        Version spirv_version = get_device_spirv_il_version(device);
+        log_info("This device supports SPIR-V offline compilation. SPIR-V version is %s\n", spirv_version.to_string().c_str());
+    }
+    return true;
+
```
Comment (aarongreig) [inline test_common/harness/testHarness.cpp:973]: extra newline

File: test_common/harness/kernelHelpers.cpp [hunk 5]
```diff
@@ -686,11 +686,32 @@ static int create_single_kernel_helper_create_program_offline(cl_context context
 
         size_t length = modifiedKernelBuf.size();
         log_info("offlineCompiler: clCreateProgramWithSource replaced with clCreateProgramWithIL\n");
+        if (gCoreILProgram) {
+            *outProgram = clCreateProgramWithIL(context, &modifiedKernelBuf[0], length, &error);
+        }
+        else {
+            cl_platform_id platform;
+            error = clGetDeviceInfo(device, CL_DEVICE_PLATFORM, sizeof(cl_platform_id), &platform, NULL);
+            print_error(error, "clGetDeviceInfo for CL_DEVICE_PLATFORM failed");
```
Comment (aarongreig) [inline test_common/harness/kernelHelpers.cpp:695]: think the result needs checking here as well

**Comment by kpet** [comment]
@gwawiork This now needs rebasing.

File: test_common/harness/kernelHelpers.cpp (hunk 5 above)
Comment (alycm) [reply test_common/harness/kernelHelpers.cpp:678]: @gwawiork I think this still needs the error code checked.

**Comment by alycm** [comment]
> I think the [linkage](https://github.com/KhronosGroup/OpenCL-CTS/blob/master/test_conformance/spirv_new/test_linkage.cpp#L36) and [opaque](https://github.com/KhronosGroup/OpenCL-CTS/blob/master/test_conformance/spirv_new/test_op_opaque.cpp#L31) tests will also need updating

@aarongreig Did these get addressed?

File: test_common/harness/kernelHelpers.cpp [hunk 6]
```diff
@@ -665,11 +665,42 @@ static int create_single_kernel_helper_create_program_offline(cl_context context
...
+        {
+            *outProgram = clCreateProgramWithIL(context, &modifiedKernelBuf[0],
+                                                length, &error);
+        }
+        else
+        {
+            cl_platform_id platform;
+            error = clGetDeviceInfo(device, CL_DEVICE_PLATFORM,
+                                    sizeof(cl_platform_id), &platform, NULL);
+            print_error(error, "clGetDeviceInfo for CL_DEVICE_PLATFORM failed");
```
Comment (aarongreig) [inline test_common/harness/kernelHelpers.cpp:678]:
```suggestion
            if(error != CL_SUCCESS)
            {
                print_error(error, "clGetDeviceInfo for CL_DEVICE_PLATFORM failed");
            }
```

**Comment by aarongreig** [comment]
> > I think the [linkage](https://github.com/KhronosGroup/OpenCL-CTS/blob/master/test_conformance/spirv_new/test_linkage.cpp#L36) and [opaque](https://github.com/KhronosGroup/OpenCL-CTS/blob/master/test_conformance/spirv_new/test_op_opaque.cpp#L31) tests will also need updating
> 
> @aarongreig Did these get addressed?

yeah that comment was addressed

File: test_common/harness/kernelHelpers.cpp (hunk 6 above)
Comment (gwawiork) [reply test_common/harness/kernelHelpers.cpp]: Ok I've used test_error which include return condition in case of error != CL_SUCCESS

**Comment by alycm** [comment]
@aarongreig Will take the version check point to a new issue/PR.

**Comment by kpet** [comment]
@alycm (and others with merge permissions): please squash when merging :). It makes the history more readable and makes it much easier to figure out the PR that the changes were made by.

**Comment by alycm** [comment]
@kpet Fair enough, will do so in the future!

**Pull request merged**